src/odemis/driver/smaract.py

```diff
@@ -3173,3 +2961,1175 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        # the functions are available in the header file. It is not clear how much time saving the
+        # parameters takes. It can be assumed that it's fast --> wait 1 s.
+        time.sleep(1)
+
+    def _wait_for_event(self, end_state, timeout=float("inf")):
+        """
+        Blocks until event is triggered or timeout.
+        end_state (int): state of event variable to wait for
+        timeout (float): maximum time to wait in s. If inf, it will wait forever.
+        """
+        if timeout == float("inf"):
+            tend = None
+        else:
+            tend = time.time() + timeout
+        ev = SA_SI_Event()
+
+        state = None
+        while state != end_state:
+            if tend:
+                t = int((tend - time.time()) * 1000)  # SA_SI_WaitForEvent accepts timeout in ms
+                if t < 0 or ev.error == SA_SI_ERROR_TIMEOUT:
+                    raise TimeoutError()
+                t = c_uint(t)
+            else:
+                t = SA_SI_TIMEOUT_INFINITE
+
+            # Don't check event type, it's not always accurate.
+            # Instead, only one event is active and all other events are disabled, so it is clear
+            # which event we are waiting for.
+            self.core.SA_SI_WaitForEvent(self._id, byref(ev), t)
+            state = ev.devEventParameter & 0xffff
+
+    @isasync
+    def reference(self, _=None):
+        """
+        This is not a "normal" referencing procedure since the hardware doesn't have any actuators. Instead,
+        it performs an internal adjustment routine which is required to get accurate position values.
+        """
+        f = self._createReferenceFuture()
+        self._executor.submitf(f, self._doReference, f)
+        return f
+
+    def _doReference(self, future):
+        """
+        Actually runs the referencing code.
+        future (Future): the future it handles
+        raise:
+            IOError: if referencing failed due to hardware
+            CancelledError if was cancelled
+        """
+        # The position polling thread can continue, it does not seem to interfere with the referencing procedure.
+        try:
+            with future._moving_lock:
+                if future._must_stop:
+                    future._was_stopped = True
+                # Reset reference so that if it fails, it states the axes are not referenced (anymore)
+                self.referenced._value = {a: False for a in self._channels.keys()}
+                logging.debug("Starting referencing.")
+                # Cannot go immediately to automatic adjustment --> first switch to manual adjustment
+                self.EnableEventNotification(SA_PS_AF_ADJUSTMENT_PROGRESS_EVENT)
+                self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_STATE_PROP,
+                                     SA_PS_ADJUSTMENT_STATE_MANUAL_ADJUST)
+            self._wait_for_event(SA_PS_ADJUSTMENT_STATE_MANUAL_ADJUST)
```

Comment:
Does it return if Cancel() was called during the wait? I guess not.... It should either return or raise a CancelledError (if there is a special event corresponding).

Reply:
It depends on what was canceled. If the event is the IS_STABLE event and I'm canceling a referencing procedure, then the waitforevent function will block forever.

Reply:
It is now working the way you requested. The problem was that I didn't raise CancelledError in enough places.